File: webapp/app/models/logistics.py
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any, Union
from datetime import datetime

# =====================================================
# 基本データモデル
# =====================================================

class LocationData(BaseModel):
    """位置情報データ"""
    name: str = Field(..., description="拠点名")
    latitude: float = Field(..., description="緯度")
    longitude: float = Field(..., description="経度")

class CustomerData(LocationData):
    """顧客データ"""
    demand: float = Field(..., description="需要量")
    weight: Optional[float] = Field(None, description="重量")

class DCData(LocationData):
    """配送センター候補データ"""
    capacity: float = Field(..., description="容量")
    fixed_cost: float = Field(..., description="固定費用")
    variable_cost: float = Field(0.0, description="変動費用")
    opening_cost: Optional[float] = Field(None, description="開設費用")

class PlantData(LocationData):
    """工場データ"""
    capacity: float = Field(..., description="生産容量")
    production_cost: float = Field(..., description="生産費用")

class ProductData(BaseModel):
    """製品データ"""
    prod_id: str = Field(..., description="製品ID")
    name: str = Field(..., description="製品名")
    weight: float = Field(..., description="重量")
    volume: float = Field(..., description="容積")
    value: float = Field(..., description="価値")
    unit_cost: float = Field(..., description="単価")

class TransportationData(BaseModel):
    """輸送データ"""
    from_location: str = Field(..., description="出発地")
    to_location: str = Field(..., description="到着地")
    distance: float = Field(..., description="距離")
    cost_per_unit: float = Field(..., description="単位輸送費用")
    transit_time: Optional[float] = Field(None, description="輸送時間")

class DemandData(BaseModel):
    """需要データ"""
    customer: str = Field(..., description="顧客名")
    product: str = Field(..., description="製品")
    period: str = Field(..., description="期間")
    demand: float = Field(..., description="需要量")

# =====================================================
# 最適化問題のリクエスト・レスポンスモデル
# =====================================================

class WeiszfeldRequest(BaseModel):
    """Weiszfeld法リクエスト"""
    customers: List[CustomerData] = Field(..., description="顧客リスト")
    num_facilities: int = Field(1, description="施設数")
    max_iterations: int = Field(1000, description="最大反復回数")
    tolerance: float = Field(1e-6, description="収束判定閾値")
    use_great_circle: bool = Field(True, description="大円距離を使用")

class WeiszfeldResult(BaseModel):
    """Weiszfeld法結果"""
    facility_locations: List[LocationData] = Field(..., description="最適施設位置")
    total_cost: float = Field(..., description="総費用")
    iterations: int = Field(..., description="実行反復回数")
    convergence_history: List[float] = Field(..., description="収束履歴")
    customer_assignments: Dict[str, int] = Field(..., description="顧客割当")

class KMedianRequest(BaseModel):
    """K-Median最適化リクエスト"""
    customers: List[CustomerData] = Field(..., description="顧客リスト")
    dc_candidates: List[DCData] = Field(..., description="DC候補")
    k: int = Field(..., description="選択するDC数")
    max_iterations: int = Field(1000, description="最大反復回数")
    learning_rate: float = Field(0.01, description="学習率")
    momentum: float = Field(0.9, description="モメンタム")

class KMedianResult(BaseModel):
    """K-Median最適化結果"""
    selected_facilities: List[int] = Field(..., description="選択された施設インデックス")
    facility_locations: List[DCData] = Field(..., description="選択された施設")
    total_cost: float = Field(..., description="総費用")
    objective_history: List[float] = Field(..., description="目的関数履歴")
    customer_assignments: Dict[str, int] = Field(..., description="顧客割当")

class ClusteringRequest(BaseModel):
    """顧客クラスタリングリクエスト"""
    customers: List[CustomerData] = Field(..., description="顧客リスト")
    method: str = Field("kmeans", description="クラスタリング手法")
    n_clusters: int = Field(..., description="クラスター数")
    use_road_distance: bool = Field(False, description="道路距離を使用")

class ClusteringResult(BaseModel):
    """クラスタリング結果"""
    clusters: Dict[str, int] = Field(..., description="顧客のクラスター割当")
    cluster_centers: List[LocationData] = Field(..., description="クラスター中心")
    aggregated_customers: List[CustomerData] = Field(..., description="集約された顧客")
    silhouette_score: float = Field(..., description="シルエット係数")

class LNDRequest(BaseModel):
    """物流ネットワーク設計リクエスト"""
    model_config = {"protected_namespaces": ()}
    
    customers: List[CustomerData] = Field(..., description="顧客リスト")
    dc_candidates: List[DCData] = Field(..., description="DC候補リスト")
    plants: List[PlantData] = Field(..., description="工場リスト")
    products: Optional[List[ProductData]] = Field(None, description="製品リスト")
    transportation_costs: Optional[List[TransportationData]] = Field(None, description="輸送費用")
    model_type: str = Field("multi_source", description="モデルタイプ")
    optimization_objective: str = Field("minimize_cost", description="最適化目標")
    capacity_constraints: bool = Field(True, description="容量制約")
    max_facilities: Optional[int] = Field(None, description="最大施設数")
    max_iterations: int = Field(500, description="最大反復回数")
    tolerance: float = Field(1e-6, description="収束判定閾値")
    co2_constraint: Optional[float] = Field(None, description="CO2制約")

class LNDResult(BaseModel):
    """物流ネットワーク設計結果"""
    selected_facilities: List[DCData] = Field(..., description="選択された施設")
    flow_assignments: Dict[str, Dict[str, float]] = Field(..., description="フロー割当")
    total_cost: float = Field(..., description="総費用")
    cost_breakdown: Dict[str, float] = Field(..., description="費用内訳")
    facility_utilization: Dict[str, float] = Field(..., description="施設稼働率")
    network_performance: Dict[str, float] = Field(..., description="ネットワーク性能")
    solution_status: str = Field(..., description="解ステータス")
    solve_time: float = Field(..., description="求解時間")
    co2_emissions: Optional[float] = Field(None, description="CO2排出量")

class NetworkVisualizationRequest(BaseModel):
    """ネットワーク可視化リクエスト"""
    lnd_result: LNDResult = Field(..., description="LND結果")
    show_flows: bool = Field(True, description="フロー表示")
    flow_threshold: float = Field(0.0, description="フロー表示閾値")
    map_style: str = Field("open-street-map", description="地図スタイル")

class NetworkVisualizationResult(BaseModel):
    """ネットワーク可視化結果"""
    plotly_figure: Dict[str, Any] = Field(..., description="Plotly図表")
    network_stats: Dict[str, float] = Field(..., description="ネットワーク統計")
    legend_data: Dict[str, Any] = Field(..., description="凡例データ")

# =====================================================
# 分析・レポートモデル
# =====================================================

class NetworkAnalysisRequest(BaseModel):
    """ネットワーク分析リクエスト"""
    lnd_result: LNDResult = Field(..., description="LND結果")
    analysis_types: List[str] = Field(..., description="分析タイプリスト")
    comparison_baseline: Optional[LNDResult] = Field(None, description="比較ベースライン")

class NetworkAnalysisResult(BaseModel):
    """ネットワーク分析結果"""
    cost_analysis: Dict[str, Any] = Field(..., description="費用分析")
    distance_analysis: Dict[str, Any] = Field(..., description="距離分析")
    efficiency_metrics: Dict[str, float] = Field(..., description="効率性指標")
    sustainability_metrics: Dict[str, float] = Field(..., description="持続可能性指標")
    risk_assessment: Dict[str, Any] = Field(..., description="リスク評価")

class ScenarioComparison(BaseModel):
    """シナリオ比較"""
    scenarios: List[LNDResult] = Field(..., description="比較シナリオ")
    comparison_metrics: List[str] = Field(..., description="比較指標")
    sensitivity_analysis: Dict[str, Any] = Field(..., description="感度分析")

# =====================================================
# Excel統合モデル
# =====================================================

class ExcelImportRequest(BaseModel):
    """Excel取込みリクエスト"""
    file_content: str = Field(..., description="Base64エンコードされたExcelファイル")
    sheet_mapping: Dict[str, str] = Field(..., description="シートマッピング")
    data_validation: bool = Field(True, description="データ検証")

class ExcelExportRequest(BaseModel):
    """Excel出力リクエスト"""
    lnd_result: LNDResult = Field(..., description="LND結果")
    template_type: str = Field("standard", description="テンプレートタイプ")
    include_charts: bool = Field(True, description="チャート含める")
    include_maps: bool = Field(False, description="地図含める")

class ExcelExportResult(BaseModel):
    """Excel出力結果"""
    file_content: str = Field(..., description="Base64エンコードされたExcelファイル")
    filename: str = Field(..., description="ファイル名")
    file_size: int = Field(..., description="ファイルサイズ")

# =====================================================
# リアルタイム最適化モデル
# =====================================================

class OptimizationProgress(BaseModel):
    """最適化進捗"""
    task_id: str = Field(..., description="タスクID")
    status: str = Field(..., description="ステータス")
    progress: float = Field(..., description="進捗率")
    current_objective: Optional[float] = Field(None, description="現在の目的関数値")
    best_solution: Optional[LNDResult] = Field(None, description="現在の最良解")
    estimated_completion: Optional[datetime] = Field(None, description="完了予想時刻")
    messages: List[str] = Field([], description="メッセージログ")

File: frontend/src/services/logistics.ts
import axios from 'axios';

const LOGISTICS_API_BASE = process.env.REACT_APP_API_URL?.replace('/api/v1', '') + '/api/v1/logistics' || 'http://localhost:8000/api/v1/logistics';

// 型定義
export interface LocationData {
  name: string;
  latitude: number;
  longitude: number;
}

export interface CustomerData extends LocationData {
  demand: number;
  weight?: number;
}

export interface DCData extends LocationData {
  capacity: number;
  fixed_cost: number;
  variable_cost: number;
  opening_cost?: number;
}

export interface PlantData extends LocationData {
  capacity: number;
  production_cost: number;
}

export interface WeiszfeldRequest {
  customers: CustomerData[];
  num_facilities: number;
  max_iterations?: number;
  tolerance?: number;
  use_great_circle?: boolean;
}

export interface WeiszfeldResult {
  facility_locations: LocationData[];
  total_cost: number;
  iterations: number;
  convergence_history: number[];
  customer_assignments: Record<string, number>;
}

export interface KMedianRequest {
  customers: CustomerData[];
  dc_candidates: DCData[];
  k: number;
  max_iterations?: number;
  learning_rate?: number;
  momentum?: number;
}

export interface KMedianResult {
  selected_facilities: number[];
  facility_locations: DCData[];
  total_cost: number;
  objective_history: number[];
  customer_assignments: Record<string, number>;
}

export interface ClusteringRequest {
  customers: CustomerData[];
  method: string;
  n_clusters: number;
  use_road_distance?: boolean;
}

export interface ClusteringResult {
  clusters: Record<string, number>;
  cluster_centers: LocationData[];
  aggregated_customers: CustomerData[];
  silhouette_score: number;
}

export interface LNDRequest {
  customers: CustomerData[];
  dc_candidates: DCData[];
  plants: PlantData[];
  model_type?: string;
  optimization_objective?: string;
  capacity_constraints?: boolean;
  max_facilities?: number;
  co2_constraint?: number;
}

export interface LNDResult {
  selected_facilities: DCData[];
  flow_assignments: Record<string, Record<string, number>>;
  total_cost: number;
  cost_breakdown: Record<string, number>;
  facility_utilization: Record<string, number>;
  network_performance: Record<string, number>;
  solution_status: string;
  solve_time: number;
  co2_emissions?: number;
}

export interface NetworkVisualizationRequest {
  lnd_result: LNDResult;
  show_flows?: boolean;
  flow_threshold?: number;
  map_style?: string;
}

export interface NetworkVisualizationResult {
  plotly_figure: any;
  network_stats: Record<string, number>;
  legend_data: Record<string, any>;
}

class LogisticsAPI {
  // ヘルスチェック
  async healthCheck(): Promise<any> {
    try {
      const response = await axios.get(`${LOGISTICS_API_BASE}/health`);
      return response.data;
    } catch (error) {
      console.error('Logistics health check failed:', error);
      throw error;
    }
  }

  // Weiszfeld法による施設立地最適化
  async optimizeFacilityLocations(request: WeiszfeldRequest): Promise<WeiszfeldResult> {
    try {
      const response = await axios.post(`${LOGISTICS_API_BASE}/weiszfeld`, request);
      return response.data;
    } catch (error) {
      console.error('Weiszfeld optimization failed:', error);
      throw error;
    }
  }

  // K-Median最適化
  async solveKMedian(request: KMedianRequest): Promise<KMedianResult> {
    try {
      const response = await axios.post(`${LOGISTICS_API_BASE}/k-median`, request);
      return response.data;
    } catch (error) {
      console.error('K-Median optimization failed:', error);
      throw error;
    }
  }

  // 顧客クラスタリング
  async clusterCustomers(request: ClusteringRequest): Promise<ClusteringResult> {
    try {
      const response = await axios.post(`${LOGISTICS_API_BASE}/clustering`, request);
      return response.data;
    } catch (error) {
      console.error('Customer clustering failed:', error);
      throw error;
    }
  }

  // 物流ネットワーク設計
  async solveLogisticsNetworkDesign(request: LNDRequest): Promise<LNDResult> {
    try {
      const response = await axios.post(`${LOGISTICS_API_BASE}/lnd`, request);
      return response.data;
    } catch (error) {
      console.error('Logistics network design failed:', error);
      throw error;
    }
  }

  // ネットワーク可視化
  async createNetworkVisualization(request: NetworkVisualizationRequest): Promise<NetworkVisualizationResult> {
    try {
      const response = await axios.post(`${LOGISTICS_API_BASE}/visualization`, request);
      return response.data;
    } catch (error) {
      console.error('Network visualization failed:', error);
      throw error;
    }
  }

  // 包括的物流分析
  async comprehensiveAnalysis(
    customers: CustomerData[],
    dc_candidates: DCData[],
    plants: PlantData[] = [],
    analysis_options?: Record<string, boolean>
  ): Promise<any> {
    try {
      const response = await axios.post(`${LOGISTICS_API_BASE}/comprehensive-analysis`, {
        customers,
        dc_candidates,
        plants,
        analysis_options
      });
      return response.data;
    } catch (error) {
      console.error('Comprehensive logistics analysis failed:', error);
      throw error;
    }
  }

  // サンプルデータ生成
  async generateSampleData(
    num_customers: number = 20,
    num_dc_candidates: number = 5,
    region: string = 'japan'
  ): Promise<any> {
    try {
      const response = await axios.post(`${LOGISTICS_API_BASE}/generate-sample-data`, {
        num_customers,
        num_dc_candidates,
        region
      });
      return response.data;
    } catch (error) {
      console.error('Sample data generation failed:', error);
      throw error;
    }
  }

  // アルゴリズムベンチマーク
  async benchmarkAlgorithms(
    customers: CustomerData[],
    dc_candidates: DCData[],
    algorithms: string[] = ['weiszfeld', 'k_median', 'basic_lnd']
  ): Promise<any> {
    try {
      const response = await axios.post(`${LOGISTICS_API_BASE}/benchmark-algorithms`, {
        customers,
        dc_candidates,
        algorithms
      });
      return response.data;
    } catch (error) {
      console.error('Algorithm benchmark failed:', error);
      throw error;
    }
  }
}

export const logisticsApi = new LogisticsAPI();